solr/core/src/java/org/apache/solr/query/FilterQuery.java

```diff
@@ -125,11 +126,24 @@ public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float bo
       return new ConstantScoreQuery(q).createWeight(searcher, scoreMode, 0f);
     }
 
-    SolrIndexSearcher solrSearcher = (SolrIndexSearcher) searcher;
-    DocSet docs = solrSearcher.getDocSet(q);
+    DocSet docs = createDocSet((SolrIndexSearcher) searcher);
     // reqInfo.addCloseHook(docs);  // needed for off-heap refcounting
 
     // note: DocSet.makeQuery is basically a CSQ
     return docs.makeQuery().createWeight(searcher, scoreMode, 0f);
   }
+
+  @Override
+  public DocSet createDocSet(SolrIndexSearcher searcher) throws IOException {
+    /*
+    There is a substantial practical reason for FilterQuery to implement DocSetProducer:
```

Comment:
Well, you know, it might not be a good idea to open this can of worms. 
There's `org.apache.solr.search.SolrIndexSearcher#getDocSetNC`, where NC implies NOcache, I suppose. That method hits  org.apache.solr.search.DocSetUtil#createDocSet(), which checks for `DocSetProducer`. I'm not sure if reaching filtercache here is what `getDocSetNC` is plan to do.

Reply:
Yes, NC=no-cache; it's a little counterintuitive on the surface, but indeed that's the fundamental _purpose_ of `FilterQuery`: it is used to consult the cache in contexts that otherwise would not cache, by reaching out to the `filterCache` "behind the scenes".

Often we would in fact expect to arrive at `FilterQuery.createDocSet()` via `SolrIndexSearcher.getDocSetNC()`, because `FilterQuery.getCacche()` now returns `false`. But what that really signifies is "don't cache the top level" ... "(because I handle reaching out the filterCache internally)". #624 fixed the fact that FilterQuery would previously have been cached twice.

This PR doesn't actually change whether `FilterQuery` hits the filterCache ... it _always_ does, by definition. It just changes whether, having hit the cache, it forces the caller to build a clone of the cached DocSet (as it currently does), or supports a way to directly return the cached DocSet instance.